data_diff/diff_tables.py

```diff
@@ -167,12 +183,12 @@ def get_values(self) -> list:
     def choose_checkpoints(self, count: int) -> List[DbKey]:
         "Suggests a bunch of evenly-spaced checkpoints to split by (not including start, end)"
         assert self.is_bounded
-        if isinstance(self.min_key, ArithUUID):
-            checkpoints = split_space(self.min_key.int, self.max_key.int, count)
-            assert isinstance(self.max_key, ArithUUID)
-            return [ArithUUID(int=i) for i in checkpoints]
 
-        return split_space(self.min_key, self.max_key, count)
+        # Split the space for each item of the key
+        splits = [split_col_space(min_i, max_i, count) for min_i, max_i in safezip(self.min_key, self.max_key)]
```

Comment:
Yes, those are serious flaws in the implementation. Thanks for catching them!

I like the idea of using only one key, but I think we should use the largest one, rather than just the first. Because, for example, you can have the `(a,b)` pair, where `a` is just 0 and 1, while `b` is the actual count.

I can still imagine some scenarios where it wouldn't be optimal, but I don't know how likely they are. I'm guessing that not very likely.

Reply:
So, one thing we learned quickly is that count can time out for really large databases. And at least for most relational databases, min/max are quite fast, which is why we're using them instead of count.

Our "gamble" is that `max-min` is a close enough approximation of the size, since we only need it to split the table evenly.

Reply:
> So, one thing we learned quickly is that count can time out for really large databases. And at least for most relational databases, min/max are quite fast, which is why we're using them instead of count.

Every one I'm familiar with has a "fast table count" query available.  They aren't part of the SQL standard, and the details of how you run it vary from DBMS to DBMS, but they work and they're accurate as long as you're not too picky about the details of how currently-in-progress transactions get counted.

Reply:
The nuance is that it is not a table count, it is a per-column count, and what's worse, it counts the distinct values of each column. So, there could be no fast hack for that. So, we should be careful about performance on big tables.

Reply:
As for the min-max approach, I'll copy it here too:

Min-max distance might be a good proxy of cardinality for numeric columns of PKs. But if that is a UUID, the min will be close to 0000-………-0000, the max will be close to perhaps FFFF-………-FFFF, so the size will be the full specter of values, even if there are 2 of them in fact (e.g. a user uuid, org uuid, etc).

Some databases prefer UUIDs as identifiers even for small dictionaries or have no auto-increment or sequence generators (see ClickHouse and essentially everything distributed with "eventual consistency" of data). So, UUIDs will be their primary PK/FK column.

So, since `count(distinct clmn)` is a problem, I suggest to do the following: 

* in this PR, only put a hint in which order to pass the PK columns in the docs/readme (the most diverse first, the least diverse last) — let's not overcomplicate it;
* create a separate feature request to implement auto-selection of the best PK column later, which can also be disabled by a config/cli option (assuming that huge databases of that scale are rare).
* create a separate feature request to do the "proper" multi-column splitting — but I have concerns about the complexity of such an algorithm and the required inputs (the column statistical profiles).

@erezsh What do you think?